rfcs/2020-07-21-2744-vicscript.md

```diff
@@ -0,0 +1,195 @@
+
+# Use slash escapes within quotes
+bar = buz."fub\"\\fob".fab
+```
+
+And supports coalescing fields within a path with a pipe operator `|`, where if the first value is null or missing then the next value is selected, and so on:
+
+```coffee
+foo = bar.(baz | buz).bev
+```
+
+You can assign a value to the root of your event by writing to the `root` keyword:
+
+```coffee
+root = nested.object
+```
+
+Vicscript supports float, int, boolean, string, null, array and object literals:
+
+```coffee
+first = 7
+second = false
+third = "string"
+fourth = null
+```
+
+As well as a timestamp type:
+
+```coffee
+# NOTE: The full API for the timestamp type is out of the scope of this RFC. The
+# following example is for demonstrative purposes.
+
+created_at = now().unix()
+# Sets `created_at` to an integer representing the current unix timestamp
+# (in seconds).
+```
+
+Boolean operators and arithmetic galore:
+
+```coffee
+is_big = number > 100
+multiplied = number * 7
+```
+
+Perform assignments conditionally with an `if` statement:
+
+```coffee
+id = id
+sorted_foo = if foo.type() == "array" { foo.sort() }
+```
+
+```text
+In:  {"id":"first","foo":"not an array"}
+Out: {"id":"first"}'
+
+In:  {"id":"second","foo":["c","a","d","b"]}
+Out: {"id":"second","sorted_foo":["a","b","c","d"]}
+```
+
+The resulting event will only contain fields that are explicitly mapped, in order to begin your map with a full copy of the input event structure you can start by assigning `this` to `root`:
+
+```coffee
+root = this
+sorted_foo = if foo.type() == "array" { foo.sort() }
+```
+
+```text
+In:  {"id":"first","foo":"not an array"}
+Out: {"id":"first","foo":"not an array"}
+
+In:  {"id":"second","foo":["c","a","d","b"]}
+Out: {"id":"second","foo":["c","a","d","b"],"sorted_foo":["a","b","c","d"]}
+```
+
+Use a wealth of methods on values in order to perform common mutations on them, since the underlying value is immutable these operations are safe to use in any combination:
+
+```coffee
+sorted = foo.sort()
+uppercase = bar.uppercase()
+```
+
+```
+In:  {"foo":["c","a","d","b"],"bar":"hello world"}
+Out: {"sorted":["a","b","c","d"],"uppercase":"HELLO WORLD"}
+```
+
+## Examples
+
+I've written an implementation of [an embedded CSV parser](2020-07-21-2744-vicscript/example1.coffee) using the above proposal in order to demonstrate what it looks like. This example is currently runnable with Benthos, and I've used the `.coffee` suffix because the syntax highlighting for CoffeeScript basically gives us everything we need already.
+
+## Prior Art
+
+### JQ
+
+JQ is a great tool and rightfully gets a lot of love. It also basically solves the same problem as Vicscript. Unfortunately, it's common knowledge that JQ doesn't score high on readability, and therefore it doesn't scale well at all in relation to mapping size and complexity (imagine trying to solve [https://github.com/timberio/vector/issues/1965](https://github.com/timberio/vector/issues/1965) with it).
```

Comment:
There are two important factors here, readability and scalability. 

Readability is a tricky one. I definitely agree that jq can be hard to understand, but that's largely a non-issue if the user is already familiar with jq from somewhere else. Given that it's pretty popular, it's hard to say how those balance out.

Scalability is one where we need to think about the target use cases. If we're expecting for this transform to handle mostly light reshaping, then scalability is less of an issue. Users will likely have the opportunity to master the basics before diving into something more complicated and may not mind spending a bit more time to figure out something they recognize is pretty complicated. On the other hand, if we expect a meaningful number of users to jump straight into larger mapping declarations, it's more important that we have something that scales smoothly to that size and level of complexity.